helix-view/src/document.rs

```diff
@@ -542,39 +581,137 @@ impl Document {
+            *current_save = None;
+            log::trace!("reawait save of '{:?}' result: {:?}", self.path(), result);
+            return Some(result);
+        }
+
+        // return early if the receiver is closed
+        let rx = self.save_receiver.as_mut()?;
+
+        let save_req = if block {
+            rx.recv().await
+        } else {
+            let msg = rx.try_recv();
+
+            if let Err(err) = msg {
+                match err {
+                    TryRecvError::Empty => return None,
+                    TryRecvError::Disconnected => None,
                 }
-                if let Some(notification) =
-                    language_server.text_document_did_save(identifier, &text)
-                {
-                    notification.await?;
+            } else {
+                msg.ok()
+            }
+        };
+
+        let save = match save_req {
+            Some(save) => save,
+            None => {
+                self.save_receiver = None;
+                return None;
+            }
+        };
+
+        // save a handle to the future so that when a poll on this
+        // function gets cancelled, we don't lose it
+        *current_save = Some(save);
+        log::trace!("awaiting save of '{:?}'", self.path());
+
+        let result = (*current_save).as_mut().unwrap().await;
+        *current_save = None;
+
+        log::trace!("save of '{:?}' result: {:?}", self.path(), result);
+
+        Some(result)
+    }
+
+    /// Flushes the queue of pending writes. If any fail,
+    /// it stops early before emptying the rest of the queue.
+    pub async fn try_flush_saves(&mut self) -> Option<DocumentSavedEventResult> {
+        self.flush_saves_impl(false).await
+    }
+
+    async fn flush_saves_impl(&mut self, block: bool) -> Option<DocumentSavedEventResult> {
+        let mut final_result = None;
+
+        while let Some(save_event) = self.await_save_impl(block).await {
+            let is_err = match &save_event {
+                Ok(event) => {
+                    self.set_last_saved_revision(event.revision);
+                    false
                 }
+                Err(err) => {
+                    log::error!(
+                        "error saving document {:?}: {}",
+                        self.path().map(|path| path.to_string_lossy()),
+                        err
+                    );
+                    true
+                }
+            };
+
+            final_result = Some(save_event);
+
+            if is_err {
+                break;
             }
+        }
+
+        final_result
+    }
 
-            Ok(())
+    /// Prepares the Document for being closed by stopping any new writes
+    /// and flushing through the queue of pending writes. If any fail,
+    /// it stops early before emptying the rest of the queue.
+    pub async fn close(&mut self) -> Option<DocumentSavedEventResult> {
+        if self.save_sender.is_some() {
+            self.save_sender.take();
         }
```

Comment:
Should be enough to just call `save_sender.take()` without the `is_some()` check